File: leetcode/0002/addTwoNumbers.swift
public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}

func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
    var l1 = l1
    var l2 = l2
    let pre = ListNode()
    var cur = pre
    var carry = 0
    
    while l1 != nil || l2 != nil {
        let x = l1 == nil ? 0 : l1!.val
        let y = l2 == nil ? 0 : l2!.val
        var sum = x + y + carry
        
        carry = sum / 10
        sum = sum % 10
        cur.next = ListNode(sum)
        
        cur = cur.next!
        if let l = l1 {
            l1 = l.next
        }
        if let l = l2 {
            l2 = l.next
        }
    }
    
    if carry == 1 {
        cur.next = ListNode(carry)
    }
    
    return pre.next
}

func test() {
    let l1 = ListNode(2)
    let l2 = ListNode(4)
    let l3 = ListNode(3)
    l1.next = l2
    l2.next = l3

    let r1 = ListNode(5)
    let r2 = ListNode(6)
    let r3 = ListNode(4)
    r1.next = r2
    r2.next = r3

    let result = addTwoNumbers(l1, r1)
    var temp = result
    print(temp?.val)
    while let _ = temp?.next {
        temp = temp?.next
        print(temp?.val)
    }
}

test()

File: leetcode/0002/addTwoNumbers.py
# -*- coding: utf-8 -*-

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def addTwoNumbers(l1: ListNode, l2: ListNode) -> ListNode:
    pre = ListNode(0)
    cur = pre
    carry = 0

    while l1 != None or l2 != None:
        x = 0 if l1 == None else l1.val
        y = 0 if l2 == None else l2.val
        sum = x + y + carry
        
        carry = int(sum / 10)
        sum = int(sum % 10)
        cur.next = ListNode(sum)

        cur = cur.next
        if l1 != None:
            l1 = l1.next
        if l2 != None:
            l2 = l2.next
    if carry == 1:
        cur.next = ListNode(carry)
    return pre.next

def test():
    l1 = ListNode(val=2)
    l2 = ListNode(val=4)
    l3 = ListNode(val=3)
    l1.next = l2
    l2.next = l3

    r1 = ListNode(val=5)
    r2 = ListNode(val=6)
    r3 = ListNode(val=4)
    r1.next = r2
    r2.next = r3

    result = addTwoNumbers(l1, r1)
    temp = result
    print(temp.val)
    while temp.next != None:
        temp = temp.next
        print(temp.val)

test()

